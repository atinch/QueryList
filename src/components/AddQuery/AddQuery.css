
.form-wrapper {
  margin: 20px auto;
  border: 2px;
  width: 600px;
}
