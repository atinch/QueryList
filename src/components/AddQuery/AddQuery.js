import React from 'react';
import uuid from 'uuid/v4';
import { connect } from 'react-redux';
import { addQueryAction } from '../../redux';
import { Form, Input, TextArea } from 'semantic-ui-react';
import { Redirect } from 'react-router-dom';

import './AddQuery.css';

  export class AddQuery extends React.Component {

  state = {
    firstName:'',
    surname: '',
    email: '',
    query: '',
    firstNameError: false,
    surnameError: false,
    emailError: false,
    redirect: false
  }

  ValidationControl = () => {

    let isFormCorrect = true
    var re = /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  
    if(!this.state.firstName.trim()) {
      this.setState({firstNameError: true})
      isFormCorrect = false;
    }else { 
      this.setState({firstNameError: false})
    }
  
    if(!this.state.surname.trim()) {
      this.setState({surnameError: true})
      isFormCorrect = false;
    }else { 
      this.setState({surnameError: false})
    }
  
    if(this.state.email.trim() && re.test(this.state.email)) {
      this.setState({emailError: false})
    }else { 
      this.setState({emailError: true})
      isFormCorrect = false;
    }
  
    return isFormCorrect
  }

   onSubmit = (event) => {

    if(!this.ValidationControl()) return false;
   
    this.setState({ redirect : true})
    event.preventDefault();
    this.props.addQueryAction({
      id: uuid(),
      firstName: this.state.firstName,
      surname: this.state.surname,
      email: this.state.email,
      query: this.state.query,
      redirect: false
    });

    this.setState({ 
      firstName: '',
      surname:'',
      email: '',
      query: ''
    })
  };

  handleChange = prop => event => {
    this.setState({ [prop]: event.target.value });
};

  render() { 
  console.log()
    const form =   (<Form onSubmit={this.onSubmit} className='form-wrapper'>
    <Form.Group widths='equal' >
        <Form.Field control={Input}
          label='First name' 
          name="firstName"
          placeholder='First name' 
          value={this.state.firstName}
          onChange={(e) => this.setState({ firstName: e.target.value}) }
          error={this.state.firstNameError}
        />
        <Form.Field control={Input}
          label="Surname"
          name="surname"
          placeholder="Surname"
          value={this.state.surname}
          onChange={(e) =>  this.setState({ surname: e.target.value})}
          error={this.state.surnameError}
        />
        <Form.Field control={Input}
          label="Email"
          name="email"
          placeholder="Email"
          value={this.state.email}
          onChange={(e) =>  this.setState({ email: e.target.value})}
          error={this.state.emailError}
        />
        </Form.Group>
        <Form.Field control={TextArea}
          label="Query"
          name="query"
          placeholder="Query"
          value={this.state.query}
          onChange={(e) =>  this.setState({ query: e.target.value})}
        />
        <div style={{textAlign: 'center'}}>
          <Form.Button type="submit">Add</Form.Button>
        </div>
    </Form>)

    return (
      <div>
        { this.props.complete && this.state.redirect ? <Redirect to='/'/> : form }
      </div>   
    );
  }
};

const mapStateToProps = (state) => ({
  queries: state.queries,
 complete: state.complete
});

export default connect(
  mapStateToProps,
  { addQueryAction }
)(AddQuery);