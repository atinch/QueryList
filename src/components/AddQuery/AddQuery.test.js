import React from 'react';
import { AddQuery } from './AddQuery';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json'
import { shallow } from 'enzyme';
import { Form, FormField } from 'semantic-ui-react';

configure({ adapter: new Adapter() });

describe('AddQuery', () => {

  const wrapper = shallow(<AddQuery/>);

  it('should render a Form', () => {
    expect(toJson(wrapper)).toMatchSnapshot()
   expect(wrapper.find(Form)).toHaveLength(1);
  });
  
});

describe('When name field is changed', () => {
  it('should change state correctly', () => {
    const wrapper = shallow(<AddQuery />)
    
    const firstName = wrapper.find(FormField).first()
    expect(firstName.prop('value')).toBe('')
    const expected = 'Sude'
    firstName.simulate("change", { target: { value: expected } });
    expect(wrapper.state('firstName')).toBe(expected)
  });
});

describe('When email field is changed', () => {
  it('should change state correctly', () => {
    const wrapper = shallow(<AddQuery />)
    
    const input = wrapper.find(FormField).at(2)
    expect(input.prop('value')).toBe('')
    expect(input.prop('name')).toBe('email')
    const expected = 'This is query...'
    input.simulate("change", { target: { value: expected } });
    expect(wrapper.state('email')).toBe(expected)
  });
});

/* describe('when form is submitted', () => {
  it('should invoke onSubmit prop with correct params', () => {

    const onSubmitMock = jest.fn()
    const props = {
      query: { id: 'sxxdfÂ£$sAS1', firstName: 'john', surname: 'mike', email: 'arc@gmail.com', query: 'What is ...' },
      onSubmit: onSubmitMock
    }
    const wrapper = shallow(<AddQuery {...props}/>)

    const expected = {
      id: 122432,
      firstName: 'Atinc',
      surname: 'Erguven',
      email: 'Atinc@gmail.com',
      query: 'Lorem ....'
    }
    
    const nameInput = wrapper.find(FormField).first()
    nameInput.simulate("change", { target: { value: expected.name } });

    const emailInput = wrapper.find(FormField).at(2)
    emailInput.simulate("change", { target: { value: expected.email } });

    wrapper.find(Form).simulate('submit', { preventDefault: jest.fn() })
    expect(onSubmitMock).toHaveBeenCalledWith(expected)
  });
}); */
