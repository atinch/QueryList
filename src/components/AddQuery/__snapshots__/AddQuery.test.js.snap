// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddQuery should be Form 1`] = `
<div>
  <Form
    as="form"
    className="form-wrapper"
    onSubmit={[Function]}
  >
    <FormGroup
      widths="equal"
    >
      <FormField
        control={[Function]}
        error={false}
        label="First name"
        name="firstName"
        onChange={[Function]}
        placeholder="First name"
        value=""
      />
      <FormField
        control={[Function]}
        error={false}
        label="Surname"
        name="surname"
        onChange={[Function]}
        placeholder="Surname"
        value=""
      />
      <FormField
        control={[Function]}
        error={false}
        label="Email"
        name="email"
        onChange={[Function]}
        placeholder="Email"
        value=""
      />
    </FormGroup>
    <FormField
      control={[Function]}
      label="Query"
      name="query"
      onChange={[Function]}
      placeholder="Query"
      value=""
    />
    <div
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      <FormButton
        as={[Function]}
        control={[Function]}
        type="submit"
      >
        Add
      </FormButton>
    </div>
  </Form>
</div>
`;

exports[`AddQuery should render a Form 1`] = `
<div>
  <Form
    as="form"
    className="form-wrapper"
    onSubmit={[Function]}
  >
    <FormGroup
      widths="equal"
    >
      <FormField
        control={[Function]}
        error={false}
        label="First name"
        name="firstName"
        onChange={[Function]}
        placeholder="First name"
        value=""
      />
      <FormField
        control={[Function]}
        error={false}
        label="Surname"
        name="surname"
        onChange={[Function]}
        placeholder="Surname"
        value=""
      />
      <FormField
        control={[Function]}
        error={false}
        label="Email"
        name="email"
        onChange={[Function]}
        placeholder="Email"
        value=""
      />
    </FormGroup>
    <FormField
      control={[Function]}
      label="Query"
      name="query"
      onChange={[Function]}
      placeholder="Query"
      value=""
    />
    <div
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      <FormButton
        as={[Function]}
        control={[Function]}
        type="submit"
      >
        Add
      </FormButton>
    </div>
  </Form>
</div>
`;
